src/Raven.Server/Documents/Handlers/BulkInsertHandler.cs

```diff
@@ -143,6 +147,46 @@ private async Task<IOperationResult> DoBulkInsert(Action<IOperationProgress> onP
             }
         }
 
+        private int? _changeVectorSize;
+
+        private long GetSize(BatchRequestParser.CommandData commandData, DocumentDatabase documentDatabase)
+        {
+            switch (commandData.Type)
+            {
+                case CommandType.PUT:
+                    return commandData.Document.Size;
+                case CommandType.Counters:
+                    long size = 0;
+                    foreach (var operation in commandData.Counters.Operations)
+                    {
+                        size += operation.CounterName.Length
+                                + sizeof(long) // etag 
+                                + sizeof(long) // counter value
+                                + GetChangeVectorSize() // estimated change vector size
+                                + 10; // estimated collection name size
+                    }
+
+                    return size;
+                default:
+                    throw new ArgumentOutOfRangeException($"'{commandData.Type}' isn't supported");
+            }
+
+
+            int GetChangeVectorSize()
+            {
+                if (_changeVectorSize.HasValue)
+                    return _changeVectorSize.Value;
+
+                using (documentDatabase.DocumentsStorage.ContextPool.AllocateOperationContext(out DocumentsOperationContext ctx))
+                using (ctx.OpenReadTransaction())
+                {
+                    var databaseChangeVector = DocumentsStorage.GetDatabaseChangeVector(ctx);
+                    _changeVectorSize = Encoding.UTF8.GetBytes(databaseChangeVector).Length;
```

Comment:
This is *really* expense. And you are allocating here.
Why do you need it? It would be sufficent to just call `changeVector.Length *2`, which should be a great estimation.

Reply:
I only call it once per the entire BulkInsert operation (the size is cached).